### ISSUE TITLE
Assignment 5

### ISSUE THREAD

**Issue by albertotrelles** [alexanderquispe/ECO224#1165]
# Instructions 

Create a new repo called DML_NN and work with your group members like in previous assignments. The repo should have the following structure 

  ```
DML_NN/
├── Python/
│   ├── output/
│   └── scripts/
├── R/
│   ├── output/
│   └── scripts/
├── Julia/
│   ├── output/
│   └── scripts/
  ```

- For visualization purposes, please provide your solutions in **Jupyter Notebooks**, including code, graphs, and explanations, using **Python, R, and Julia**.  Don't forget to export graphs to the /output folder. 
- You may choose to place everything in a **single notebook** or split the work across **multiple notebooks**, as long as the submissions are in **Jupyter Notebook format** 
- If you have trouble with Jupyter Notebooks you can use VScode or some other IDE, just present things in an orderly manner and clearly respond to questions as comments when required. Also export tables to the /output folder if you won't present your results in Jupyter Notebooks.
- You can load the data by calling the link (see question 2). However, if you load locally without the link please include an input folder with the data in it. 

# Question 1 - NN Basics

## I. Fitting Data (8 points)

Simulate the following DGP: $y = \text{sin}(x) + \varepsilon$
You can consider $x \in [0, 2\pi]$

Next, train a NN to predict $y$ with 3 hidden layers, each with 50 neurons. Try the following activation functions:
- logistic
- tanh
- relu

Finally, train another NN with the same characteristics but using a **different activation function for each hidden layer** (you can leave any other parameters as default when training the NN). 

For each NN trained, plot the data and the fitted values in the XY-plane
Which NN fits the data better? 

*Points*
- Simulation: 0.5
- NN training: 2
- Graphs: 2
- Question: 0.5 

## II. Learning-rate 

Briefly explain what the learning-rate is in the context of Neural Networks.

Use the same data simulation and the activation function that performed better between logistic, tanh, and relu. Iterate over the following learning rates training for each iteration a NN of a single hidden layer with 50 neurons.

- 0.0001, 0.001, 0.01, 0.1 

In one single graph, plot the data and the fitted values in the XY-plane. Properly label the graph and answer which learning rate worked best. 

Repeat this process (training and graph) for a NN with 2 hidden layers, and 3 hidden layers (each layer with 50 neurons). In this illustrative example, what can you say about the relationship between the learning rate and the number of hidden layers? 

*Points*
- NN training for every lr: 1.5
- Graphs: 1
- Question: 0.5 

# Question 2 - DML (12 points)

Use the <code>penn_jae.dat</code> dataset available [here](https://raw.githubusercontent.com/VC2015/DMLonGitHub/master/penn_jae.dat). 

## I. Cleaning and set-up

- Keep only those observations where 'tg' is equal to 0 and 4 
- Define a dummy variable 'T4' equal to 1 whenever 'tg' equals 4 (this will be our treatment variable $d$)
- Define the outcome variable $y$ as the natural log of 'inuidur1'.
- Create dummy variables 'dep_0', 'dep_1, 'dep_2' for the possible categories of the original variable 'dep'
- Define $x$ to include the following features 
    - 'female','black','othrace'
    - 'dep_1','dep_2'
    - 'q2','q3','q4','q5','q6'
    - 'recall', 'agelt35','agegt54',
    - 'durable', 'nondurable', 'lusd','husd'

*Points*
- Loading data and cleaning: 0.5
- Setting $y,d,x$ for estimation: 1

## II. Debiased ML 

- Just like in the labs, define your dml() function for the Partially Linear Model setting with cross-fitting. 
- Next, estimate using DML the effect of $d$ on $y$. Present your results in a table like in the labs. 
- Based on your results, pick which model(s) to use and estimate using the selected models. 

Use OLS with controls, Lasso, Random Forests, and a NN of your choosing.  

*Points*
- Function: 1 
- OLS, Lasso, RF: 2
- NN: 2
- Table: 0.5
- Model selection and estimation: 0.5

## III. No cross-fitting 

- Now define a new function that operates just like dml() but without cross-fitting for estimating $\tilde{y}$ and $\tilde{d}$
- Repeat the same process as with dml() (estimate the causal effect using your new function, present the results, and select and estimate a final model)

Finally, compare your results from dml() and your new function that doesn't use cross-fitting and answer the following. 
- What can you say about the RMSE for predicting $y$ and $d$? 
- Why is it that estimating with one function yields lower RMSE than another? 
- What problem would we have if we chose to estimate without cross-fitting? 

*Points*
- Function: 1
- OLS, Lasso, RF, NN: 1
- Table: 0.5
- Model selection and estimation: 0.5
- Questions: 1.5
